docs/guide/mocking.md

```diff
@@ -1,293 +1,274 @@
 
-For example:
+const getLatest = (index = messages.items.length - 1) => messages.items[index]
 
-```ts
-const mockFn = vi.fn().mockImplementation(apples => apples + 1);
-// or: vi.fn(apples => apples + 1);
+const messages = {
+  items: [
+    { message: 'Simple test message', from: 'Testman' },
+    // ...
+  ],
+  getLatest, // can also be a `getter or setter if supported`
+}
 
-const NelliesBucket = mockFn(0);
-const BobsBucket = mockFn(1);
+describe('reading messages', () => {
+  afterEach(() => {
+    vi.restoreAllMocks()
+  })
 
-NelliesBucket === 1; // true
-BobsBucket === 2; // true
+  it('should get the latest message with a spy', () => {
+    const spy = vi.spyOn(messages, 'getLatest')
+    expect(spy.getMockName()).toEqual('getLatest')
 
-mockFn.mock.calls[0][0] === 0; // true
-mockFn.mock.calls[1][0] === 1; // true
-```
+    expect(messages.getLatest()).toEqual(
+      messages.items[messages.items.length - 1]
+    );
 
-### mockImplementationOnce
+    expect(spy).toHaveBeenCalledTimes(1)
 
-**Type:** `(fn: Function) => MockInstance`
+    spy.mockImplementationOnce(() => 'access-restricted')
+    expect(messages.getLatest()).toEqual('access-restricted')
 
-Accepts a function that will be used as an implementation of the mock for one call to the mocked function. Can be chained so that multiple function calls produce different results.
+    expect(spy).toHaveBeenCalledTimes(2)
+  });
 
-```ts
-const myMockFn = vi
-  .fn()
-  .mockImplementationOnce(() => true)
-  .mockImplementationOnce(() => false);
+  it('should get with a mock', () => {
+    const mock = vi.fn().mockImplementation(getLatest)
 
-myMockFn(); // true
-myMockFn(); // false
-```
+    expect(mock()).toEqual(messages.items[messages.items.length - 1])
+    expect(mock).toHaveBeenCalledTimes(1)
 
-When the mocked function runs out of implementations, it will invoke the default implementation that was set with `vi.fn(() => defaultValue)` or `.mockImplementation(() => defaultValue)` if they were called:
+    mock.mockImplementationOnce(() => 'access-restricted')
+    expect(mock()).toEqual('access-restricted')
 
-```ts
-const myMockFn = vi
-  .fn(() => 'default')
-  .mockImplementationOnce(() => 'first call')
-  .mockImplementationOnce(() => 'second call');
+    expect(mock).toHaveBeenCalledTimes(2)
 
-// 'first call', 'second call', 'default', 'default'
-console.log(myMockFn(), myMockFn(), myMockFn(), myMockFn());
+    expect(mock()).toEqual(messages.items[messages.items.length - 1])
+    expect(mock).toHaveBeenCalledTimes(3)
+  })
+})
 ```
 
-### mockReturnThis
+### More
 
-**Type:** `() => MockInstance`
+- [Jest's Mock Functions](https://jestjs.io/docs/mock-function-api)
 
-Sets inner implementation to return `this` context.
+## Modules
 
-### mockReturnValue
+Mock modules observe third-party-libraries, that are invoked in some other code, allowing you to test arguments, output or even redeclare its implementation.
```

Comment:
Can we have more of a usecase-based explanation as to why someone would want to do this?

Reply:
I was struggling a bit coming up with an example that wouldn't be too framework specific, but would still be very use-cased focused.  I didn't want to go the obvious route of axios or fetch (as we do not recommend `mocking` those in the next section 😁 ). A router example would be next on my list but would also contain a lot of framework (or even library specific) API in the example.

So I'm open to suggestions here (if anyone has any) 🙂

Reply:
What about something like mocking a DB library? Or an authorized SDK? Something where mocking just the request isn't sufficient (I think of like Firebase's Firestore which needs a separate emulator and auth to work and the network requests it makes are opaque; you could just mock the module instead to test data in and out)

Reply:
I've updated the example and went for the db one but I do think this is now the only example that won't run out of the box (a.k.a copy pasted into vite.dev/new) and I'm not sure if this makes things clearer, but it is a more usecase focussed example. Let me know what you think 👌